cmd/buildkitd/debug_flight.go

```diff
@@ -0,0 +1,113 @@
+}
+
+func newFlightRecorder() *flightRecorder {
+	dbg := &flightRecorder{
+		recorder: trace.NewFlightRecorder(),
+	}
+	dbg.cond = sync.NewCond(&dbg.mu)
+	return dbg
+}
+
+func (r *flightRecorder) StartTrace(w http.ResponseWriter, req *http.Request) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.recorder.Enabled() {
+		http.Error(w, "flight recorder is already running", http.StatusConflict)
+		return
+	}
+	if err := r.recorder.Start(); err != nil {
+		http.Error(w, fmt.Sprintf("could not start flight recorder: %s", err), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (r *flightRecorder) StopTrace(w http.ResponseWriter, req *http.Request) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.recorder.Enabled() {
+		http.Error(w, "flight recorder is not running", http.StatusConflict)
+		return
+	}
+	if err := r.recorder.Stop(); err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func (r *flightRecorder) SetTracePeriod(w http.ResponseWriter, req *http.Request) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if r.recorder.Enabled() {
+		http.Error(w, "flight recorder is running, stop it to change its period", http.StatusPreconditionFailed)
+		return
+	}
+	periodValue := req.FormValue("period")
+	period, err := time.ParseDuration(periodValue)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("invalid flight recorder period: %s", err), http.StatusBadRequest)
+	}
+	r.recorder.SetPeriod(period)
+}
+
+func (r *flightRecorder) Trace(w http.ResponseWriter, req *http.Request) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="trace"`)
+	if _, err := r.recorder.WriteTo(w); err != nil {
+		http.Error(w, fmt.Sprintf("could not write in-flight trace: %s", err), http.StatusInternalServerError)
+	}
+}
+
+func (r *flightRecorder) StartCPUProfile(w http.ResponseWriter, req *http.Request) {
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Disposition", `attachment; filename="profile"`)
+
+	if err := pprof.StartCPUProfile(w); err != nil {
+		http.Error(w, fmt.Sprintf("could not start CPU profile: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	r.mu.Lock()
+	r.cond.Wait()
+	r.mu.Unlock()
+}
+
+func (r *flightRecorder) StopCPUProfile(w http.ResponseWriter, req *http.Request) {
+	pprof.StopCPUProfile()
+	r.cond.Broadcast()
+}
+
+func setupDebugFlight(m *http.ServeMux) {
+	r := newFlightRecorder()
+
+	const (
+		flightPattern      = "/debug/flight"
+		flightTracePattern = flightPattern + "/trace"
+		flightCPUPattern   = flightPattern + "/cpu"
+	)
+
+	m.HandleFunc("POST "+flightTracePattern+"/start", r.StartTrace)
+	m.HandleFunc("POST "+flightTracePattern+"/stop", r.StopTrace)
+	m.HandleFunc("POST "+flightTracePattern+"/set_period", r.SetTracePeriod)
+	m.HandleFunc("GET "+flightTracePattern, r.Trace)
+
+	m.HandleFunc("POST "+flightCPUPattern+"/start", r.StartCPUProfile)
```

Comment:
These are already available in `/debug/pprof/profile`

Reply:
The default does seem safer. If we want to keep the new ones then:
- Why do these make sense under `/flight` as it is just plain CPU profile.
- `StartCPUProfile` should stop the profiling when connection drops and request context gets cancelled.

Reply:
I think it might be better to remove the CPU profiler endpoints from this PR and focus on the trace recorder. We can add the CPU profilers in a subsequent PR. WDYT?

Reply:
sgtm

Reply:
I've left only the trace profiler. PTAL.